### PR TITLE
Support traversables inside `in` operator

### PR DESCRIPTION
**Pull request by ostrolucky**

### CODE REVIEW

File: Visitor/Asserter.php [hunk 1]
```diff
@@ -89,7 +89,15 @@ public function __construct(Ruler\Context $context = null)
         $this->setOperator('>=',  function ($a, $b) { return $a >= $b; });
         $this->setOperator('<',   function ($a, $b) { return $a <  $b; });
         $this->setOperator('<=',  function ($a, $b) { return $a <= $b; });
-        $this->setOperator('in',  function ($a, array $b) { return in_array($a, $b); });
+        $this->setOperator('in',  function ($a, $b) {
+            foreach ($b as $element) {
+                if ($a == $element) {
```
Comment (Pierozi) [inline Visitor/Asserter.php]: Could you use strict equality please.

**Review by Pierozi — DISMISSED** [review]
Using foreach looks fine to me, in this case the performance should be similar.

File: Visitor/Asserter.php (hunk 1 above)
Comment (ostrolucky) [reply Visitor/Asserter.php]: I can, but previous in_array construct didn't use strict check either, so this will be BC break. Do you still want it?

Comment (Pierozi) [reply Visitor/Asserter.php]: Good point, but guess it should have been strict in first place, @Hywan, @vonglasow thoughts?

Comment (ostrolucky) [reply Visitor/Asserter.php]: IMHO non-strict check is better fit here. 

- No BC break. 
- Consistent with symfony expression language. 
- Easier to understand for users of your DSL (they don't understand concept of types, they won't understand why `'1'` doesn't match `1`). 
- Keeps working with enum objects and cloned objects. That's common pitfall with strict checks..

File: Visitor/Asserter.php [hunk 2]
```diff
@@ -89,7 +89,15 @@ public function __construct(Ruler\Context $context = null)
         $this->setOperator('>=',  function ($a, $b) { return $a >= $b; });
         $this->setOperator('<',   function ($a, $b) { return $a <  $b; });
         $this->setOperator('<=',  function ($a, $b) { return $a <= $b; });
-        $this->setOperator('in',  function ($a, array $b) { return in_array($a, $b); });
+        $this->setOperator('in',  function ($a, $b) {
```
Comment (stephpy) [inline Visitor/Asserter.php]: IMHO we still have to make a check:

```
if(false === is_array( $b ) && !$b instanceof Traversable ) {
  throw new \InvalidArgumentException(....
}
```

File: Visitor/Asserter.php (hunk 1 above)
Comment (stephpy) [reply Visitor/Asserter.php]: To keep BC, i'm agree.
Anyway, if somebody want a strict check, he just has to create a new operator.

Comment (stephpy) [reply Visitor/Asserter.php]: @Pierozi unfortunately PHP `in_array` function is not strict at all.

```
var_dump(in_array(0, ['foo'])); // true
```

Comment (ostrolucky) [reply Visitor/Asserter.php]: Here's my stackoverflow question regarding this https://stackoverflow.com/questions/43276251/make-in-array-not-convert-between-numeric-and-alpha-values/48083328#48083328. Might be worthwile to replace current algorithm with array_intersect + iterator_to_array combo

File: Visitor/Asserter.php (hunk 2 above)
Comment (ostrolucky) [reply Visitor/Asserter.php]: Well similar error would be produced by PHP automatically during foreach call, no need to do it ourselves. But will do anyway.

**Review by jubianchi — APPROVED** [review]

File: Test/Unit/Visitor/Asserter.php [hunk 3]
```diff
@@ -308,12 +308,18 @@ public function case_operator_in_empty_array()
 
     public function case_operator_in()
     {
-        return $this->_case_boolean_operator('in', [7, [1, 3, 5, 7, 9]], true);
+        $this->_case_boolean_operator('in', [7, [1, 3, 5, 7, 9]], true);
```
Comment (Hywan) [inline Test/Unit/Visitor/Asserter.php]: This test case now contains 2 test cases. Please, split it :-).

File: Test/Unit/Visitor/Asserter.php [hunk 4]
```diff
@@ -308,12 +308,18 @@ public function case_operator_in_empty_array()
...
+
+        $generator = function () { yield 1; yield 7;};
+        $this->_case_boolean_operator('in', [7, $generator()], true);
     }
 
     public function case_operator_in_falsy()
     {
-        return $this->_case_boolean_operator('in', [42, [1, 3, 5, 7, 9]], false);
+        $this->_case_boolean_operator('in', [42, [1, 3, 5, 7, 9]], false);
```
Comment (Hywan) [inline Test/Unit/Visitor/Asserter.php]: Same, please split the test case.

File: Visitor/Asserter.php [hunk 5]
```diff
@@ -89,7 +89,19 @@ public function __construct(Ruler\Context $context = null)
...
+            if (false === is_array($b) && !$b instanceof \Traversable) {
+                throw new \InvalidArgumentException(sprintf('Expect iterable, got %s', get_class($b)));
```
Comment (Hywan) [inline Visitor/Asserter.php]: Please throw a [`Hoa\Ruler\Exception\Asserter`](https://github.com/hoaproject/Ruler/blob/master/Exception/Asserter.php) instance.

File: Visitor/Asserter.php [hunk 6]
```diff
@@ -89,7 +89,19 @@ public function __construct(Ruler\Context $context = null)
         $this->setOperator('>=',  function ($a, $b) { return $a >= $b; });
         $this->setOperator('<',   function ($a, $b) { return $a <  $b; });
         $this->setOperator('<=',  function ($a, $b) { return $a <= $b; });
-        $this->setOperator('in',  function ($a, array $b) { return in_array($a, $b); });
+        $this->setOperator('in',  function ($a, $b) {
```
Comment (Hywan) [inline Visitor/Asserter.php]: Why not using the `iterable` type hint?

**Review by Hywan — CHANGES_REQUESTED** [review]

**Comment by Hywan** [comment]
Thanks for the PR!

File: Visitor/Asserter.php (hunk 6 above)
Comment (ostrolucky) [reply Visitor/Asserter.php]: you support PHP 5.6

File: Test/Unit/Visitor/Asserter.php (hunk 3 above)
Comment (ostrolucky) [reply Test/Unit/Visitor/Asserter.php]: It's ok if test case contains multiple asserts. But ok will split.

File: Visitor/Asserter.php (hunk 6 above)
Comment (Hywan) [reply Visitor/Asserter.php]: We are dropping PHP 5.x support. So go on :-).

Comment (ostrolucky) [reply Visitor/Asserter.php]: Ok but iterable typehint is available since PHP 7.1 only. So can I drop all other versions from travis matrix?

**Comment by ostrolucky** [comment]
@Hywan No problem! So, how should we go forward with this? Do you want me to typehint to iterable and drop support for PHP < 7.1, or just change the type of exception thrown?

**Comment by Hywan** [comment]
@ostrolucky We are discussing about merging the `php7` branch into `master`,  so that's it's going to be much easier for you. Sorry for the late reply, I'm back from my paternity leave :-).

**Comment by ostrolucky** [comment]
All I see on gh is master branch, no php7?

**Comment by Hywan** [comment]
`php7` has been merged inside `master`.

**Pull request closed without merging**